packages/hydrogen/src/product/getProductOptions.ts

```diff
@@ -0,0 +1,215 @@
+ * {
+ *    '[0,0]': {id: 1, selectedOptions: [{name: 'Color', value: 'Red'}, {name: 'Size', value: 'Small'}]},
+ *    '[0,1]': {id: 2, selectedOptions: [{name: 'Color', value: 'Red'}, {name: 'Size', value: 'Medium'}]},
+ * }
+ * @param variants
+ * @param productOptionMappings
+ * @returns
+ */
+function mapAdjacentVariants(variants: ProductVariant[], productOptionMappings: ProductOptionsMapping[]) {
+  return Object.assign({}, ...variants.map((variant) => {
+    const variantKey = encodeSelectedProductOptionAsKey(variant.selectedOptions || [], productOptionMappings);
+    return {[variantKey]: variant};;
+  }));
+}
+
+type MappedProductOptionValue = ProductOptionValue & {
+  variant: ProductVariant;
+  handle: string;
+  variantUriQuery: string;
+  selected: boolean;
+  exists: boolean;
+  available: boolean;
+  isDifferentProduct: boolean;
+}
+
+export type MappedProductOptions = Omit<ProductOption, 'optionValues'> & {
+  optionValues: MappedProductOptionValue[];
+}
+
+export function getProductOptions(product: Product): MappedProductOptions[] {
+  const {
+    options,
+    //@ts-ignore
+    selectedOrFirstAvailableVariant: selectedVariant,
+    //@ts-ignore
+    adjacentVariants,
+    //@ts-ignore
+    encodedVariantExistence,
+    //@ts-ignore
+    encodedVariantAvailability,
+    handle: productHandle,
+  } = product;
+  // Get a mapping of product option names to their index for matching encoded values
+  const productOptionMappings = mapProductOptions(options);
+
+  // Get the adjacent variants mapped to the encoded selected option values
+  const variants = mapAdjacentVariants([
+    selectedVariant,
+    ...adjacentVariants,
+  ], productOptionMappings);
+
+  // Get the key:value version of selected options for building url query params
+  const selectedOptions = mapSelectedProductOptionToObject(selectedVariant.selectedOptions || []);
+
+  return options.map((option, optionIndex) => {
+    return {
+      ...option,
+      optionValues: option.optionValues.map((value) => {
+        const targetOptionParams = {...selectedOptions};  // Clones the selected options
+
+        // Modify the selected option value to the current option value
+        targetOptionParams[option.name] = value.name;
+
+        // Encode the new selected option values as a key for mapping to the product variants
+        const targetKey = encodeSelectedProductOptionAsKey(targetOptionParams || [], productOptionMappings);
+
+        // Top-down option check for existence and availability
+        const topDownKey = JSON.parse(targetKey).slice(0, optionIndex + 1);
+        const exists = isOptionValueInEncoding(topDownKey, encodedVariantExistence);
+        const available = isOptionValueInEncoding(topDownKey, encodedVariantAvailability);
+
+        // Get the variant for the current option value if exists, else use the first selectable variant
+        // @ts-ignore
+        const variant: ProductVariant = variants[targetKey] || value.firstSelectableVariant;
```

Comment:
I think this should be:

```suggestion
        const variant: ProductVariant = variants[targetKey] || product.selectedOrFirstAvailableVariant;
```

Reply:
It's not. `product.options.optionValues.firstSelectableVariant` is the first available variant of this product variant. For example ...

We have product options as follows and the selected option is ['Red', 'Small'] and colors options is a combined listing.
 
```
[
  ['Red', 'Blue'],
  ['Small', 'Medium', 'Large]
]
```

Let's say I need to generate a link for `Blue` option, however, ['Blue', 'Small'] is an option that doesn't exist and the first available variant is ['Blue', 'Medium'].

This is where we need the  `product.options.optionValues.firstSelectableVariant` which will return the variant for ['Blue', 'Medium']